{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "23ecf57a-a77e-4dce-a3af-9e4e8d42c829",
   "metadata": {},
   "source": [
    "#### Important user-defined variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e99a4cf-0dc5-478e-b96a-65fa065149d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "SPECIES_NAME=Chrysophrys_auratus\n",
    "SPECIES_SYMBOL=ChrAur  # Just a shortened name of your species, for naming files and genes\n",
    "GENOME=/workspace/cfnjxb/PB_assembly/00_Assembly_V2/ChrAur2-renamed.scaffold.fasta\n",
    "LIBRARY=/workspace/cfnjxb/PB_assembly/02_repeats/intermediate/ChrAur_scaff-families.fa\n",
    "LEFT_READ=/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/InputData/combined_snapper_trimmed_reads_left.fq.gz\n",
    "RIGHT_READ=/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/InputData/combined_snapper_trimmed_reads_right.fq.gz\n",
    "LIFTOFF_REF_GFF=/workspace/cfnjxb/PB_assembly/01_annotations/intermediate_files/GCF_904848185.1_fAcaLat1.1_genomic.gff\n",
    "LIFTOFF_REF_GENOME=/workspace/cfnjxb/PB_assembly/01_annotations/intermediate_files/GCF_904848185.1_fAcaLat1.1_genomic.fna\n",
    "SPECIES=/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/InputData/species.tab  # Tab separated file of species names and their orthodb name\n",
    "# ODB_LINEAGE=https://bioinf.uni-greifswald.de/bioinf/partitioned_odb12/Vertebrata.fa.gz  # Currently doesn't exist. Code has been modified so this isn't needed\n",
    "EMAIL=sam.modern@plantandfood.co.nz\n",
    "WORKSPACE_TO_BIND_1=/workspace/cfnsjm  # This should be the workspace you're running this in"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18ce9c68-a53d-4ec0-a924-f1ec3d8bcc17",
   "metadata": {},
   "source": [
    "#### Variables you might want to change:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee51479a-89d6-4ffe-b528-099e2d41d3a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE_DIR=$PWD\n",
    "ODB_VERSION_NUMBER=\"12\"\n",
    "BUSCO_LINEAGE=actinopterygii_odb10\n",
    "KMER_LEN=13\n",
    "LONG_TIMEOUT=\"4-00:00:00\"\n",
    "MEDIUM_TIMEOUT=\"2-00:00:00\"\n",
    "SHORT_TIMEOUT=\"04:00:00\"\n",
    "LARGE_MEM='80G'\n",
    "MEDIUM_MEM='32G'\n",
    "SMALL_MEM='16G'\n",
    "CPUS=8\n",
    "WORKSPACE_TO_BIND_2=/workspace/cfnjxb\n",
    "TE_TOOLS=/workspace/cfnjxb/dfam-tetools.sh\n",
    "SRC=$BASE_DIR/src\n",
    "LOG=$BASE_DIR/Log\n",
    "INPUT_DATA=$BASE_DIR/InputData\n",
    "FINAL_OUTPUT=$BASE_DIR/FinalOutput"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b4b3072-5732-42ba-82be-b463ae7458e4",
   "metadata": {},
   "source": [
    "## A: BRAKER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9cc2dce1-1734-4ba4-963c-359bdd512737",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORKING_D=$BASE_DIR/Braker\n",
    "REPMASK=$WORKING_D/01_Repmask\n",
    "STAR=$WORKING_D/02_Star\n",
    "BRAKER=$WORKING_D/03_Braker\n",
    "BUSCO=$WORKING_D/04_Busco\n",
    "LIFTOFF=$WORKING_D/05_Liftoff\n",
    "MERGE_ANNOTATION=$WORKING_D/06_Merge\n",
    "TMP=$WORKING_D/tmp\n",
    "OUTPUT=$WORKIND_D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fade62b0-e227-487a-9a24-18a5a7cb3129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "** INFO ** : singularity has been deprecated - please use apptainer in place.\n",
      "Loading \u001b[1mapptainer/1.1\u001b[22m\n",
      "  \u001b[91mERROR\u001b[0m: Module cannot be loaded due to a conflict.\n",
      "    HINT: Might try \"module unload singularity\" first.\n",
      "\n",
      "Loading \u001b[1mBUSCO/v5.8.2\u001b[22m\n",
      "  \u001b[91mERROR\u001b[0m: Load of requirement apptainer failed\n"
     ]
    }
   ],
   "source": [
    "ml conda\n",
    "conda deactivate\n",
    "export TERM=rxvt\n",
    "ml singularity\n",
    "ml BUSCO\n",
    "ml RepeatMasker\n",
    "ml Slurm\n",
    "ml STAR\n",
    "export TERM=dumb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6a0daad5-843f-42ff-9cc1-fb0f6733596a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/FinalOutput’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Log’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker/03_Braker’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker/04_Busco’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker/05_Liftoff’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker/06_Merge’: File exists\n"
     ]
    }
   ],
   "source": [
    "mkdir $FINAL_OUTPUT\n",
    "mkdir $WORKING_D\n",
    "mkdir $LOG\n",
    "mkdir $REPMASK\n",
    "mkdir $BRAKER\n",
    "mkdir $STAR\n",
    "mkdir $BUSCO\n",
    "mkdir $LIFTOFF\n",
    "mkdir $MERGE_ANNOTATION\n",
    "mkdir $TMP\n",
    "mkdir $TMP/singularity-tmp\n",
    "mkdir $TMP/singularity-cache\n",
    "export SINGULARITY_TMPDIR=$TMP/singularity-tmp\n",
    "export SINGULARITY_CACHEDIR=$TMP/singularity-cache"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "352f068d-aa04-40bb-839a-d31c91d5578f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sbatch --time=$SHORT_TIMEOUT --mem=$SMALL_MEM --cpus-per-task=1 --job-name=$USER\" BREAKER_build\" \\\n",
    "    --output=$LOG\"/BRAKER_build-%A.out\" --mail-type=ALL --mail-user=$EMAIL \\\n",
    "    --wrap=\"singularity build braker3.sif docker://teambraker/braker3:latest\"\n",
    "# These lines are currently commented out because the ODB partition hasn't been updated yet\n",
    "# Currently the code is set up to use a pre-generated Vertebrata orthodb partition located in my workspace\n",
    "#wget -P $INPUT_DATA $ODB_LINEAGE\n",
    "#gzip -d $INPUT_DATA/Vertebrata.fa.gz\n",
    "# Remove these lines when the proper ODB partition is updates:\n",
    "vertebrata=/workspace/cfnsjm/my_files/C_auratus/genomics/OrthoDB/orthodb-clades/clades/Vertebrata.fa\n",
    "ln -s $vertebrata $INPUT_DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7ef2aa2a-5e38-4eff-822f-774d4b51dba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv braker3.sif $SRC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aabdc97a-a36d-476a-9c4d-4a1b82537284",
   "metadata": {},
   "source": [
    "#### Rename sequences in the genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cdac672f-8d0c-47c1-b5ec-fb5d0e772950",
   "metadata": {},
   "outputs": [],
   "source": [
    "python $SRC/rename_fasta.py $GENOME $INPUT_DATA/renamed_genome.fasta $INPUT_DATA/genome_renamed_mappings.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab21cc4-8c94-4aa8-96f1-1000f32311e1",
   "metadata": {},
   "source": [
    "#### 01-RepMask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ccdefdb5-5815-4cde-ad5b-05c971e163a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 8735832\n"
     ]
    }
   ],
   "source": [
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH --time=$MEDIUM_TIMEOUT\n",
    "#SBATCH --cpus-per-task=$CPUS\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --job-name=$USER\" RepMask\"\n",
    "#SBATCH --mem=$SMALL_MEM\n",
    "#SBATCH --output=$LOG\"/01_Masker-%A.out\"\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --mail-user=$EMAIL\n",
    "\n",
    "$TE_TOOLS -- RepeatMasker \"$INPUT_DATA/renamed_genome.fasta\" -lib \"$LIBRARY\" -pa $CPUS -nolow -xsmall -dir=\"$REPMASK\"\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f07f8d87-3ff2-44a9-be5b-ddbf451fdf36",
   "metadata": {},
   "source": [
    "#### 02-STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6860ca6f-6ca4-4282-9cb9-b1c4aea02b25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 7767146\n"
     ]
    }
   ],
   "source": [
    "mkdir $STAR/IndexedGenome\n",
    "\n",
    "sbatch --time=$SHORT_TIMEOUT --mem=$LARGE_MEM --cpus-per-task=$CPUS --job-name=$USER\" STAR Index\" --output=$LOG\"/02_STAR1-%A.out\" \\\n",
    "    --mail-type=ALL --mail-user=$EMAIL \\\n",
    "    --wrap=\"STAR --runMode genomeGenerate --runThreadN=$CPUS --genomeDir=$STAR/IndexedGenome --genomeFastaFiles=$INPUT_DATA/renamed_genome.fasta --genomeSAindexNbases=$KMER_LEN\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4df0e9b4-7277-41cb-aa0e-d053b616c961",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 7767151\n"
     ]
    }
   ],
   "source": [
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH --time=$SHORT_TIMEOUT\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --mem=$LARGE_MEM\n",
    "#SBATCH --job-name=$USER\" STAR\"\n",
    "#SBATCH --cpus-per-task=$CPUS\n",
    "#SBATCH --output=$LOG\"/02_STAR2-%A.out\"\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --mail-user=$EMAIL\n",
    "\n",
    "STAR --runThreadN=$CPUS --genomeDir=$STAR/IndexedGenome \\\n",
    "     --readFilesIn $LEFT_READ $RIGHT_READ \\\n",
    "     --outFileNamePrefix $STAR/ --outSAMtype BAM Unsorted --outSAMstrandField \\\n",
    "     intronMotif --readFilesCommand zcat\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4c6f92c-5c15-4567-a6c5-b437c11eafd1",
   "metadata": {},
   "source": [
    "#### 03-BRAKER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ba2b262f-460a-4797-8d19-3dd422e69a6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 7767272\n"
     ]
    }
   ],
   "source": [
    "mkdir $BRAKER/AugustusConfig\n",
    "singularity exec --bind=$WORKSPACE_TO_BIND_1,$WORKSPACE_TO_BIND_2 \\\n",
    "    $SRC/braker3.sif cp -a /opt/Augustus/config/. $BRAKER/AugustusConfig/\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH --time=$LONG_TIMEOUT\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --cpus-per-task=$CPUS\n",
    "#SBATCH --job-name=$USER\" BRAKER\"\n",
    "#SBATCH --mem=$LARGE_MEM\n",
    "#SBATCH --output=$LOG\"/03_BRAKER-%A.out\"\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --mail-user=$EMAIL\n",
    "\n",
    "cd /home\n",
    "\n",
    "singularity exec --bind=$WORKSPACE_TO_BIND_1,$WORKSPACE_TO_BIND_2 \\\n",
    "    $SRC/braker3.sif braker.pl --workingdir=$BRAKER --genome=$REPMASK/renamed_genome.fasta.masked --bam=$STAR/Aligned.out.bam  \\\n",
    "        --prot_seq=$INPUT_DATA/Vertebrata.fa --threads=$CPUS --species=$SPECIES_SYBMOL --AUGUSTUS_CONFIG_PATH=$BRAKER/AugustusConfig/ \\\n",
    "        --busco_lineage=$BUSCO_LINEAGE --gff3\n",
    "        \n",
    "cd -\n",
    "\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad013c79-3074-45b5-ae98-c6608e93ce2d",
   "metadata": {},
   "source": [
    "#### 04-BUSCO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "debe5e9e-fc0c-4aeb-b7dc-7ada5e7307f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-11-02 10:38:47 INFO:\tDownloading information on latest versions of BUSCO data...\n",
      "2024-11-02 10:38:50 INFO:\tDownloading file 'https://busco-data.ezlab.org/v5/data/lineages/actinopterygii_odb10.2024-01-08.tar.gz'\n",
      "2024-11-02 10:39:09 INFO:\tDecompressing file '/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/busco_downloads/lineages/actinopterygii_odb10.tar.gz'\n"
     ]
    }
   ],
   "source": [
    "mkdir $BUSCO/Database\n",
    "busco --download $BUSCO_LINEAGE\n",
    "mv busco_downloads/* $BUSCO/Database\n",
    "rm -r busco_downloads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "42941a6e-1d3f-4832-be1c-ef9823e4f1d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 7770420\n",
      "Submitted batch job 7770421\n"
     ]
    }
   ],
   "source": [
    "if test -f $BRAKER/bbc/better.gtf; then\n",
    "  sbatch << EOF\n",
    "    #!/bin/bash -e\n",
    "    #SBATCH --time=$SHORT_TIMEOUT\n",
    "    #SBATCH --ntasks-per-node=1\n",
    "    #SBATCH --cpus-per-task=$CPUS\n",
    "    #SBATCH --job-name=$USER\" BUSCO\"\n",
    "    #SBATCH --mem=$SMALL_MEM\n",
    "    #SBATCH --output=$LOG\"/04_BUSCO3-%A.out\"\n",
    "    #SBATCH --mail-type=ALL\n",
    "    #SBATCH --mail-user=$EMAIL\n",
    "    \n",
    "    busco --mode proteins -i $BRAKER/bbc/better.aa --out_path $BUSCO/ -o bbc -l $BUSCO_LINEAGE \\\n",
    "        --offline --download_path $BUSCO/Database --cpu $CPUS\n",
    "EOF\n",
    "fi\n",
    "\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH --time=$SHORT_TIMEOUT\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --cpus-per-task=$CPUS\n",
    "#SBATCH --job-name=$USER\" BUSCO\"\n",
    "#SBATCH --mem=$SMALL_MEM\n",
    "#SBATCH --output=$LOG\"/04_BUSCO1-%A.out\"\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --mail-user=$EMAIL\n",
    "\n",
    "busco --mode proteins -i $BRAKER/braker.aa --out_path $BUSCO/ -o braker -l $BUSCO_LINEAGE \\\n",
    "    --offline --download_path $BUSCO/Database --cpu $CPUS\n",
    "EOF\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH --time=$SHORT_TIMEOUT\n",
    "#SBATCH --ntasks-per-node=1\n",
    "#SBATCH --cpus-per-task=$CPUS\n",
    "#SBATCH --job-name=$USER\" BUSCO\"\n",
    "#SBATCH --mem=$SMALL_MEM\n",
    "#SBATCH --output=$LOG\"/04_BUSCO2-%A.out\"\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --mail-user=$EMAIL\n",
    "\n",
    "busco --mode proteins -i $BRAKER/Augustus/augustus.hints.aa --out_path $BUSCO/ -o augustus -l $BUSCO_LINEAGE \\\n",
    "    --offline --download_path $BUSCO/Database --cpu $CPUS\n",
    "EOF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13fd0949-176a-4765-9da6-33e3984cb50f",
   "metadata": {},
   "source": [
    "#### 05-Liftoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d8fb956f-2aaf-44f4-b630-00119ee777ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfnjxb_liftoff) \n",
      "(/workspace/appscratch/miniconda/cfnjxb_liftoff) \n",
      "Submitted batch job 7768791\n",
      "(/workspace/appscratch/miniconda/cfnjxb_liftoff) \n",
      "(/workspace/appscratch/miniconda/cfnjxb_liftoff) \n"
     ]
    }
   ],
   "source": [
    "conda activate cfnjxb_liftoff\n",
    "\n",
    "sbatch << EOF\n",
    "#!/bin/bash -e\n",
    "#SBATCH -J $USER\" Liftoff\"\t#job name\n",
    "#SBATCH -o $LOG\"/05_Liftoff-%A.out\"\t#outfile\n",
    "#SBATCH --mail-user $EMAIL\t#email when finished\n",
    "#SBATCH --mail-type=ALL\t#email settings\n",
    "#SBATCH --time=$LONG_TIMEOUT\t#booking time\n",
    "#SBATCH -n 1\t#number of processors\n",
    "#SBATCH --cpus-per-task=$CPUS\n",
    "#SBATCH --mem=$MEDIUM_MEM\n",
    "\n",
    "liftoff -o $LIFTOFF/liftoff.gff -exclude_partial -polish -p $CPUS -g $LIFTOFF_REF_GFF $GENOME $LIFTOFF_REF_GENOME\n",
    "EOF\n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c75001-f645-4f0d-b9fa-f337e5444fd1",
   "metadata": {},
   "source": [
    "#### 06-MergeAnnotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e3651480-0a08-4ce7-b460-22791b394601",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "Submitted batch job 7771102\n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n"
     ]
    }
   ],
   "source": [
    "conda activate cfnsjm_general\n",
    "braker_gtf=$BRAKER/braker.gtf\n",
    "if test -f $BRAKER/bbc/better.gtf; then\n",
    "    braker_gtf=$BRAKER/bbc/better.gtf\n",
    "fi\n",
    "\n",
    "sbatch --time=$SHORT_TIMEOUT --mem=$SMALL_MEM --cpus-per-task=1 --job-name=$USER\" Merge\" --output=$LOG\"/06_Merge-%A.out\" \\\n",
    "    --mail-type=ALL --mail-user=$EMAIL \\\n",
    "    --wrap=\"python $SRC/01_merge_annotation.py $GENOME $INPUT_DATA/genome_renamed_mappings.tsv $LIFTOFF/liftoff.gff_polished $braker_gtf $MERGE_ANNOTATION/\"\n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e69d3e8d-bb50-42ab-b22f-eff6d543a66d",
   "metadata": {},
   "source": [
    "## B: OrthoFinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1ad32944-1312-4872-8253-04a3b1f344dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "WORKING_D=$BASE_DIR/OrthoFinder\n",
    "ORTHODB_PARTITION=$INPUT_DATA/Vertebrata.fa\n",
    "PROTEOME=$MERGE_ANNOTATION/merged_ok.pep\n",
    "ODB_VERSION=odb${ODB_VERSION_NUMBER}v0\n",
    "ORTHOFINDER=$WORKING_D/08_OrthoFinder\n",
    "ADD_NAMES=$WORKING_D/09_AddGeneNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bcba8cf8-a0eb-4900-a3cd-c6e4b4684469",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda deactivate\n",
    "ml singularity\n",
    "ml orthofinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "39b342c5-a52d-4387-8931-d918a5f83fa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/OrthoFinder’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/InputData/Proteomes’: File exists\n",
      "mkdir: cannot create directory ‘/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/OrthoFinder/AddGeneNames’: File exists\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir $WORKING_D\n",
    "mkdir $INPUT_DATA/Proteomes\n",
    "mkdir $ADD_NAMES\n",
    "mkdir $ORTHOFINDER"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ca31a2a-a620-4c17-8a29-e1d166670193",
   "metadata": {},
   "source": [
    "#### 07-Download and extract data (this step can cause issues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "00afdc1d-df9e-4610-bdc9-d507b9ec2ee9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-11-03 13:48:26--  https://data.orthodb.org/v12/download/odb12v0_gene_xrefs.tab.gz\n",
      "Resolving data.orthodb.org (data.orthodb.org)... 129.194.190.40\n",
      "Connecting to data.orthodb.org (data.orthodb.org)|129.194.190.40|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4692118032 (4.4G) [application/octet-stream]\n",
      "Saving to: ‘odb12v0_gene_xrefs.tab.gz’\n",
      "\n",
      "odb12v0_gene_xrefs. 100%[===================>]   4.37G  10.9MB/s    in 6m 57s  \n",
      "\n",
      "2024-11-03 13:55:26 (10.7 MB/s) - ‘odb12v0_gene_xrefs.tab.gz’ saved [4692118032/4692118032]\n",
      "\n",
      "--2024-11-03 13:55:26--  https://zfin.org/downloads/uniprot.txt\n",
      "Resolving zfin.org (zfin.org)... 184.171.92.30\n",
      "Connecting to zfin.org (zfin.org)|184.171.92.30|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1937501 (1.8M) [text/plain]\n",
      "Saving to: ‘uniprot.txt.1’\n",
      "\n",
      "uniprot.txt.1       100%[===================>]   1.85M  1.91MB/s    in 1.0s    \n",
      "\n",
      "2024-11-03 13:55:28 (1.91 MB/s) - ‘uniprot.txt.1’ saved [1937501/1937501]\n",
      "\n",
      "--2024-11-03 13:55:28--  https://data.orthodb.org/v12/download/odb12v0_species.tab.gz\n",
      "Resolving data.orthodb.org (data.orthodb.org)... 129.194.190.40\n",
      "Connecting to data.orthodb.org (data.orthodb.org)|129.194.190.40|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 644375 (629K) [application/octet-stream]\n",
      "Saving to: ‘odb12v0_species.tab.gz’\n",
      "\n",
      "odb12v0_species.tab 100%[===================>] 629.27K   458KB/s    in 1.4s    \n",
      "\n",
      "2024-11-03 13:55:31 (458 KB/s) - ‘odb12v0_species.tab.gz’ saved [644375/644375]\n",
      "\n",
      "--2024-11-03 13:55:31--  ftp://ftp.ncbi.nih.gov/gene/DATA/gene_info.gz\n",
      "           => ‘gene_info.gz’\n",
      "Resolving ftp.ncbi.nih.gov (ftp.ncbi.nih.gov)... 130.14.250.31, 130.14.250.7, 130.14.250.11, ...\n",
      "Connecting to ftp.ncbi.nih.gov (ftp.ncbi.nih.gov)|130.14.250.31|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /gene/DATA ... done.\n",
      "==> SIZE gene_info.gz ... 1239313418\n",
      "==> PASV ... done.    ==> RETR gene_info.gz ... done.\n",
      "Length: 1239313418 (1.2G) (unauthoritative)\n",
      "\n",
      "gene_info.gz        100%[===================>]   1.15G  19.7MB/s    in 62s     \n",
      "\n",
      "2024-11-03 13:56:37 (19.1 MB/s) - ‘gene_info.gz’ saved [1239313418]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd $INPUT_DATA\n",
    "wget https://data.orthodb.org/v$ODB_VERSION_NUMBER/download/${ODB_VERSION}_gene_xrefs.tab.gz\n",
    "wget https://zfin.org/downloads/uniprot.txt\n",
    "wget https://data.orthodb.org/v$ODB_VERSION_NUMBER/download/${ODB_VERSION}_species.tab.gz\n",
    "wget 'ftp://ftp.ncbi.nih.gov/gene/DATA/gene_info.gz'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be65866d-8a38-4efa-9bd6-feb71ea92eba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "gzip: odb12v0_gene_xrefs.tab.gz: decompression OK, trailing garbage ignored\n",
      "mv: '/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/InputData/species.tab' and 'species.tab' are the same file\n",
      "srun: job 7771104 queued and waiting for resources\n",
      "srun: job 7771104 has been allocated resources\n"
     ]
    }
   ],
   "source": [
    "gzip -d ${ODB_VERSION}_gene_xrefs.tab.gz\n",
    "gzip -d ${ODB_VERSION}_species.tab.gz\n",
    "mv $SPECIES species.tab\n",
    "gzip -d gene_info.gz\n",
    "mv gene_info gene_info.tab\n",
    "cut -f 2 species.tab > taxa_ids.txt\n",
    "sed -i -e 's/^/\\^/' taxa_ids.txt\n",
    "srun --mem='16G' --time='01:00:00' grep -f taxa_ids.txt ${ODB_VERSION}_gene_xrefs.tab > gene_xrefs.tab\n",
    "grep 'NCBIgid$' gene_xrefs.tab > gene_ncbi_ids.tab\n",
    "cut -d '_' -f 1 taxa_ids.txt > taxa_ids.tab\n",
    "mv taxa_ids.tab taxa_ids.txt\n",
    "sed -i 's/$/\\t/' taxa_ids.txt\n",
    "grep -f taxa_ids.txt gene_info.tab | awk -F'\\t' '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$9\"\\t\"$10}' > subset_gene_info.tab\n",
    "sort -k2,2 subset_gene_info.tab > subset_gene_info_sorted.tab\n",
    "sort -k2,2 gene_ncbi_ids.tab > gene_ncbi_ids_sorted.tab\n",
    "head -n 1 gene_info.tab | awk -F'\\t' '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$9\"\\t\"$10\"\\todb_gene\\tignore\"}' > joined_gene_info.tab\n",
    "join -1 2 -2 2 -t $'\\t' subset_gene_info_sorted.tab gene_ncbi_ids_sorted.tab >> joined_gene_info.tab\n",
    "rm gene_info.tab\n",
    "rm taxa_ids.txt\n",
    "rm ${ODB_VERSION}_gene_xrefs.tab\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9c5160e0-4658-4662-b7a4-58e18bd110ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 7772077\n"
     ]
    }
   ],
   "source": [
    "cp $PROTEOME $INPUT_DATA/Proteomes/$SPECIES_NAME.pep\n",
    "sbatch --time=$SHORT_TIMEOUT --mem=$MEDIUM_MEM --job-name=$USER\" 07_ExtractProteomes\" \\\n",
    "    --mail-type=ALL --mail-user=$EMAIL --output=$LOG\"/07_Proteomes-%A.out\" \\\n",
    "    --wrap=\"python $SRC/extract_proteomes.py $ORTHODB_PARTITION $INPUT_DATA/species.tab $INPUT_DATA/Proteomes\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ffd8989-3f00-4435-91d5-bb5ba503cbe0",
   "metadata": {},
   "source": [
    "#### 08-OrthoFinder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d6aedb23-4ca1-406a-a847-a757cff5edb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 7772151\n"
     ]
    }
   ],
   "source": [
    "sbatch --time=$MEDIUM_TIMEOUT --mem=$MEDIUM_MEM --cpus-per-task=$CPUS --job-name=$USER\" 08_OrthoFinder\" \\\n",
    "    --output=$LOG\"/08_OrthoFinder-%A.out\" --mail-type=ALL --mail-user=$EMAIL\\\n",
    "    --wrap=\"orthofinder -t $CPUS -f $INPUT_DATA/Proteomes/ -o $ORTHOFINDER/Results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "92b5c8c7-6624-47ec-97d4-6f0dfd4f12be",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv $ORTHOFINDER/Results/* $FINAL_OUTPUT/OrthoFinder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccc65ef-9ebf-41ee-a456-ea7bf02d5d54",
   "metadata": {},
   "source": [
    "#### 09-Add gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dca6c1fa-8244-43fe-a3ca-0b1ff9c438d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "Submitted batch job 8486276\n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n",
      "(/workspace/appscratch/miniconda/cfnsjm_general) \n"
     ]
    }
   ],
   "source": [
    "conda activate cfnsjm_general\n",
    "\n",
    "sbatch --time=$MEDIUM_TIMEOUT --mem=$SMALL_MEM --cpus-per-task=1 --job-name=$USER\" Names\" --output=$LOG\"/09_Names-%A.out\" \\\n",
    "    --mail-type=ALL --mail-user=$EMAIL \\\n",
    "    --wrap=\"python $SRC/temp.py $MERGE_ANNOTATION/merged.gff $INPUT_DATA/joined_gene_info.tab $FINAL_OUTPUT/OrthoFinder/ \\\n",
    "    $ADD_NAMES/ ${SPECIES_SYMBOL}.gff $SPECIES_SYMBOL\"\n",
    "\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70829744-5e92-492b-8219-a8558d2d3cb1",
   "metadata": {},
   "source": [
    "## Finalise!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c954c344-18d0-4bc8-9320-06cdfb52d1c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv $ADD_NAMES/${SPECIES_SYMBOL}.gff $FINAL_OUTPUT\n",
    "mv $ADD_NAMES/orthogroups_with_names.tsv $FINAL_OUTPUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "82388150-c6ae-4d8f-9732-ebc7b41b9bc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm -r $REPMASK\n",
    "rm -r $STAR\n",
    "rm -r $TMP\n",
    "rm -r $ORTHOFINDER\n",
    "rm -r $BUSCO/Database\n",
    "rmdir $ADD_NAMES\n",
    "rmdir $WORKING_D"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f738e5-491a-4ee3-96f0-fa2ae58e1757",
   "metadata": {},
   "source": [
    "## Utility Blocks:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "818240a9-8797-4ec6-9043-26f53e6b5975",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JobID           JobName  Partition      State  Timelimit ExitCode               Start                 End  MaxVMSize  AllocCPUS     ReqMem AllocNodes   NTasks \n",
      "------------ ---------- ---------- ---------- ---------- -------- ------------------- ------------------- ---------- ---------- ---------- ---------- -------- \n",
      "8571905      Kraken2_c+      short OUT_OF_ME+   20:00:00    0:125 2025-05-09T14:40:28 2025-05-09T14:44:56                    16        64G          1          \n",
      "8571905.bat+      batch            OUT_OF_ME+               0:125 2025-05-09T14:40:28 2025-05-09T14:44:56          0         16                     1        1 \n",
      "8571905.ext+     extern             COMPLETED                 0:0 2025-05-09T14:40:28 2025-05-09T14:44:56          0         16                     1        1 \n"
     ]
    }
   ],
   "source": [
    "sacct --format=jobid,jobname,partition,state,TimeLimit,exitcode,start,end,maxvmsize,alloccpus,ReqMem,allocnodes,ntasks -j 8571905"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "27b04e7c-8cc0-4070-bad3-92f01062f877",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/workspace/cfnjxb/PB_assembly/00_Assembly_V2/ChrAur2-renamed.scaffold.fasta\n",
      "/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/InputData/genome_renamed_mappings.tsv\n",
      "/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker/05_Liftoff/liftoff.gff_polished\n",
      "/powerplant/workspace/cfnsjm/my_files/C_auratus/genomics/Samnotation/Braker/03_Braker/braker.gtf\n"
     ]
    }
   ],
   "source": [
    "echo $GENOME\n",
    "echo $INPUT_DATA/genome_renamed_mappings.tsv \n",
    "echo $LIFTOFF/liftoff.gff_polished \n",
    "echo $BRAKER/braker.gtf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
